contracts/margin/external/exchangewrappers/ZeroExV2MultiOrderExchangeWrapper.sol

```diff
@@ -0,0 +1,428 @@
+                order.takerAssetAmount,
+                order.makerAssetAmount
+            );
+
+            // bound the remaining amounts by the maker amount still needed
+            if (remaining.makerAmount > running.makerAmount) {
+                remaining.makerAmount = running.makerAmount;
+                remaining.takerAmount = MathHelpers.getPartialAmountRoundedUp(
+                    order.takerAssetAmount,
+                    order.makerAssetAmount,
+                    remaining.makerAmount
+                );
+            }
+
+            // update the running tallies
+            running.takerAmount = running.takerAmount.add(remaining.takerAmount);
+            running.makerAmount = running.makerAmount.sub(remaining.makerAmount);
+        }
+
+        // require that all amount was bought
+        require(
+            running.makerAmount == 0,
+            "ZeroExV2MultiOrderExchangeWrapper#getExchangeCost: Cannot buy enough maker token"
+        );
+
+        // validate that max price will not be violated
+        validateTradePrice(
+            priceRatio,
+            running.takerAmount,
+            desiredMakerToken
+        );
+
+        // return the amount of taker token needed
+        return running.takerAmount;
+    }
+
+    // ============ Private Functions ============
+
+    /**
+     * Validates that a certain takerAmount and makerAmount are within the maxPrice bounds
+     */
+    function validateTradePrice(
+        TokenAmounts memory priceRatio,
+        uint256 takerAmount,
+        uint256 makerAmount
+    )
+        private
+        pure
+    {
+        require(
+            priceRatio.makerAmount == 0 ||
+            takerAmount.mul(priceRatio.makerAmount) <= makerAmount.mul(priceRatio.takerAmount),
+            "ZeroExV2MultiOrderExchangeWrapper#validateTradePrice: Price greater than maxPrice"
+        );
+    }
+
+    /**
+     * Validates that there is at least one order in orderData and that the length is correct
+     */
+    function validateOrderData(
+        bytes memory orderData
+    )
+        private
+        pure
+    {
+        require(
+            orderData.length >= PRICE_DATA_LENGTH + ORDER_DATA_LENGTH
+            && orderData.length.sub(PRICE_DATA_LENGTH) % ORDER_DATA_LENGTH == 0,
+            "ZeroExV2MultiOrderExchangeWrapper#parseOrder: Invalid orderData length"
+        );
+    }
+
+    /**
+     * Returns the number of orders given in the orderData
+     */
+    function parseNumOrders(
+        bytes memory orderData
+    )
+        private
+        pure
+        returns (uint256)
+    {
+        return orderData.length.sub(PRICE_DATA_LENGTH).div(ORDER_DATA_LENGTH);
+    }
+
+    /**
+     * Gets the bit-offset of the index'th order in orderData
+     */
+    function getOrderDataOffset(
+        uint256 index
+    )
+        private
+        pure
+        returns (uint256)
+    {
+        return PRICE_DATA_LENGTH + index * ORDER_DATA_LENGTH;
```

Comment:
I feel like we should just use safemath here. I know you'd run out of gas long before you could pass in enough bytes to overflow a uint256, but it's just better not to have to worry about those kind of assumptions affecting things